OpenRA.Game/Activities/Activity.cs

```diff
@@ -31,45 +31,56 @@ public abstract class Activity
 	{
 		public ActivityState State { get; private set; }
 
-		Activity childActivity;
-		protected Activity ChildActivity
-		{
-			get { return childActivity != null && childActivity.State < ActivityState.Done ? childActivity : null; }
-			set { childActivity = value; }
-		}
-
+		protected Activity ChildActivity { get; private set; }
 		public Activity NextActivity { get; protected set; }
 
 		public bool IsInterruptible { get; protected set; }
+		public bool ChildHasPriority { get; protected set; }
 		public bool IsCanceling { get { return State == ActivityState.Canceling; } }
 
 		public Activity()
 		{
 			IsInterruptible = true;
+			ChildHasPriority = true;
 		}
 
 		public Activity TickOuter(Actor self)
 		{
-			if (State == ActivityState.Done && Game.Settings.Debug.StrictActivityChecking)
-				throw new InvalidOperationException("Actor {0} attempted to tick activity {1} after it had already completed.".F(self, this.GetType()));
-
 			if (State == ActivityState.Queued)
 			{
 				OnFirstRun(self);
 				State = ActivityState.Active;
 			}
 
+			if (ChildHasPriority)
+			{
+				ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
+				if (ChildActivity != null)
+					return this;
+			}
+
 			var ret = Tick(self);
-			if (ret != this)
+
+			if (ChildActivity != null && (ChildActivity.State == ActivityState.Queued || !ChildHasPriority))
```

Comment:
Can you please add a comment here explaining that we want to tick the child activity if it was just queued to avoid a single-tick delay where the actor appears to stall without doing anything?

Reply:
I assume that we have already reached the stage where we would never not want to pre-tick the child, so that it is fine to not expose a way to not do this.

Reply:
Done.

All instances where we did not pretick a childactivity where because it was already queued in `Activity.OnFirstRun`, whenever it was queued in `Activity.Tick` we did do a pre-tick. So automating this pattern by letting `TickOuter` figure it out makes sense to me.